File: benchmark/python/listmatrix.py
"""List benchmark matrix — Python peer for mfb's list.mfb plain groups.

Mirrors mfb list (Fixed) / list (Dynamic) row for row (see benchmark/mfb/
gen_list.py). Same sizes for both element types; only the element differs
(Integer vs "s"+i String). C/Python carry only the plain element axis — the
Record/State variants are an mfb value-semantics (bug-430) story. Replaces the
old `list` + `liststr` groups. Most checksums are count/sum based; the sort
adaptivity rows use an order-sensitive polynomial hash identical to mfb.
"""
import sys

RUN = 1
now_ns = None
record = None

GROUPS = [("Fixed", "lf", "int"), ("Dynamic", "ld", "str")]


def el(ty, i):
    return i if ty == "int" else "s" + str(i)


def rng(ty, n):
    return [el(ty, i) for i in range(n)]


def _pad6(n):
    s = str(n)
    return "0" * (6 - len(s)) + s if len(s) < 6 else s


def _sorthash_int(a):
    acc = 0
    for v in a:
        acc = (acc * 31 + v) % 1000000007
    return acc


def _sorthash_str(a):
    acc = 0
    for v in a:
        acc = (acc * 31 + len(v)) % 1000000007
    return acc


def run_all(run, now_ns_fn, record_fn):
    global RUN, now_ns, record
    RUN, now_ns, record = run, now_ns_fn, record_fn
    for label, pfx, ty in GROUPS:
        group = "list (%s)" % label
        is_int = ty == "int"

        def emit(op, fn):
            times = []
            checksum = 0
            for _ in range(RUN):
                t0 = now_ns()
                checksum = fn()
                times.append(now_ns() - t0)
            print("test_%s_%s = %d" % (pfx, op, checksum), file=sys.stderr)
            record(group, op, times)

        # predicates / helpers per element type
        if is_int:
            isPos = lambda n: n > 0
            isEven = lambda n: n % 2 == 0
            geHi = lambda n: n >= 999
            leLo = lambda n: n <= 5
            addf = lambda a, n: a + n
            transf = lambda n: n + n
            sortkey = lambda n: -n
            gkey = lambda n: n % 100
            r_init = 0
            r_len = lambda v: v            # reduce result agg
            g_len = lambda v: v            # get/find agg
        else:
            isPos = lambda s: len(s) > 0        # strNonEmpty (all pass)
            isEven = lambda s: len(s) <= 2      # strIsShort (filter)
            geHi = lambda s: len(s) >= 4        # strIsLong (findIndex)
            leLo = lambda s: len(s) <= 2        # strIsShort (findLastIndex)
            addf = lambda a, s: a + s
            transf = lambda s: "[" + s + "]"
            sortkey = lambda s: len(s)
            gkey = lambda s: len(s)
            r_init = ""
            r_len = lambda v: len(v)
            g_len = lambda v: len(v)

        pos = [el(ty, i) for i in range(1, 1001)] if is_int else rng(ty, 1000)
        neg = [-(i) for i in range(1, 1001)] if is_int else rng(ty, 1000)
        base1k = rng(ty, 1000)

        # --- accumulate-from-empty ---------------------------------------
        def op_append():
            xs = []
            for i in range(1000):
                xs.append(el(ty, i))
            return len(xs)
        emit("append", op_append)

        ten = list(range(10)) if is_int else ["a" + str(i) for i in range(10)]

        def op_append_batch():
            xs = []
            for _ in range(100):
                xs.extend(ten)
            return len(xs)
        emit("append_batch", op_append_batch)

        def op_prepend():
            xs = []
            for i in range(1000):
                xs.insert(0, el(ty, i))
            return len(xs)
        emit("prepend", op_prepend)

        def op_insert():
            xs = []
            for i in range(1000):
                xs.insert(len(xs) // 2, el(ty, i))
            return len(xs)
        emit("insert", op_insert)

        # --- read-only ---------------------------------------------------
        emit("copy", lambda: sum(len(list(base1k)) for _ in range(1000)))

        dup = [el(ty, i % 1000) for i in range(5000)]
        emit("distinct", lambda: len(list(dict.fromkeys(dup))))

        g2000 = rng(ty, 2000)

        def op_groupby():
            g = {}
            for v in g2000:
                g.setdefault(gkey(v), []).append(v)
            return len(g)
        emit("groupby", op_groupby)

        def op_set():
            xs = rng(ty, 200)
            for _ in range(10):
                for j in range(200):
                    v = xs[j]
                    xs[j] = v + 1 if is_int else v + "!"
            return sum((xs[j] if is_int else len(xs[j])) for j in range(200))
        emit("set", op_set)

        rand50 = rng(ty, 50)          # deterministic; sort checksum is len
        emit("sort", lambda: len(sorted(rand50)))

        asc = list(range(20000)) if is_int else [_pad6(i) for i in range(20000)]
        desc = list(range(19999, -1, -1)) if is_int else [_pad6(19999 - i) for i in range(20000)]
        scr = [(i * 7919) % 20000 for i in range(20000)] if is_int else [_pad6((i * 7919) % 20000) for i in range(20000)]
        sh = _sorthash_int if is_int else _sorthash_str
        emit("sort_asc", lambda: sh(sorted(asc)))
        emit("sort_desc", lambda: sh(sorted(desc)))
        emit("sort_rand", lambda: sh(sorted(scr)))

        emit("all", lambda: sum(1 for _ in range(200) if all(isPos(v) for v in pos)))
        emit("any", lambda: sum(1 for _ in range(200) if not any(isPos(v) for v in neg)))
        emit("chunks", lambda: sum(len([base1k[i:i + 10] for i in range(0, len(base1k), 10)]) for _ in range(200)))
        target_contains = 1000 if is_int else "s1000"
        emit("contains", lambda: sum(1 for _ in range(500) if target_contains not in base1k))
        emit("drop", lambda: sum(len(base1k[500:]) for _ in range(500)))
        emit("filter", lambda: sum(len([v for v in base1k if isEven(v)]) for _ in range(200)))
        find_t = 999 if is_int else "s999"

        def _find(xs, t):
            try:
                return xs.index(t)
            except ValueError:
                return -1
        emit("find", lambda: sum(_find(base1k, find_t) for _ in range(500)))

        def _findidx(xs, p):
            for i, v in enumerate(xs):
                if p(v):
                    return i
            return -1
        emit("findIndex", lambda: sum(_findidx(base1k, geHi) for _ in range(500)))

        def _findlast(xs, p):
            for i in range(len(xs) - 1, -1, -1):
                if p(xs[i]):
                    return i
            return -1
        emit("findLastIndex", lambda: sum(_findlast(base1k, leLo) for _ in range(500)))

        nested = [rng(ty, 10) for _ in range(100)]
        emit("flatten", lambda: sum(len([x for row in nested for x in row]) for _ in range(200)))

        def op_foreach():
            acc = 0
            for _ in range(200):
                for v in base1k:
                    acc += v if is_int else len(v)
            return acc
        emit("forEach", op_foreach)

        emit("get", lambda: sum(g_len(base1k[i]) for _ in range(100) for i in range(1000)))
        emit("getOr", lambda: sum(g_len(base1k[i]) for _ in range(100) for i in range(1000)))
        emit("mid", lambda: sum(len(base1k[250:750]) for _ in range(500)))

        def op_partition():
            acc = 0
            for _ in range(200):
                m = [v for v in base1k if isEven(v)]
                acc += len(m)
            return acc
        emit("partition", op_partition)

        def _reduce(xs):
            a = r_init
            for v in xs:
                a = addf(a, v)
            return a
        emit("reduce", lambda: sum(r_len(_reduce(base1k)) for _ in range(500)))
        emit("reduceRight", lambda: sum(r_len(_reduce(list(reversed(base1k)))) for _ in range(500)))

        def op_removeat():
            xs = list(base1k)
            cnt = 0
            while xs:
                xs.pop(0)
                cnt += 1
            return cnt
        emit("removeAt", op_removeat)

        rep_a = 500 if is_int else "s5"
        rep_b = 500 if is_int else "S5"
        emit("replace", lambda: sum(len([rep_b if v == rep_a else v for v in base1k]) for _ in range(200)))

        base500 = rng(ty, 500)
        emit("sortBy", lambda: sum(g_len(sorted(base500, key=sortkey)[0]) for _ in range(200)))

        if is_int:
            emit("sum", lambda: sum(sum(base1k) for _ in range(1000)))

        emit("take", lambda: sum(len(base1k[:500]) for _ in range(500)))
        emit("transform", lambda: sum(len([transf(v) for v in base1k]) for _ in range(200)))
        emit("window", lambda: sum(len([base1k[i:i + 10] for i in range(len(base1k) - 9)]) for _ in range(100)))
        emit("zip", lambda: sum(len(list(zip(base1k, base1k))) for _ in range(100)))

File: benchmark/c/listmatrix.c
/* GROUP: list matrix — Fixed/Dynamic element.
 * C peer for mfb's list.mfb plain groups (see benchmark/mfb/gen_list.py). Same
 * sizes for both element types; only the element differs (int vs "s"+i). C/Python
 * carry only the plain element axis — Record/State are an mfb value-semantics
 * (bug-430) story. Replaces the old `list` + `liststr` groups. Most checksums are
 * count/sum based; the sort adaptivity rows use the same polynomial hash as mfb. */
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "bench.h"
#include "listmatrix.h"

/* ---- growable lists -------------------------------------------------- */
typedef struct { long long *a; int n, cap; } IL;
static IL *il_new(int cap) { if (cap < 1) cap = 1; IL *l = malloc(sizeof(IL)); l->a = malloc(sizeof(long long) * cap); l->n = 0; l->cap = cap; return l; }
static void il_push(IL *l, long long v) { if (l->n == l->cap) { l->cap *= 2; l->a = realloc(l->a, sizeof(long long) * l->cap); } l->a[l->n++] = v; }
static void il_ins(IL *l, int idx, long long v) { if (l->n == l->cap) { l->cap *= 2; l->a = realloc(l->a, sizeof(long long) * l->cap); } memmove(&l->a[idx + 1], &l->a[idx], sizeof(long long) * (l->n - idx)); l->a[idx] = v; l->n++; }
static void il_free(IL *l) { free(l->a); free(l); }

typedef struct { char **a; int n, cap; } SL;
static SL *sl_new(int cap) { if (cap < 1) cap = 1; SL *l = malloc(sizeof(SL)); l->a = malloc(sizeof(char *) * cap); l->n = 0; l->cap = cap; return l; }
static void sl_push(SL *l, const char *s) { if (l->n == l->cap) { l->cap *= 2; l->a = realloc(l->a, sizeof(char *) * l->cap); } l->a[l->n++] = strdup(s); }
static void sl_ins(SL *l, int idx, const char *s) { if (l->n == l->cap) { l->cap *= 2; l->a = realloc(l->a, sizeof(char *) * l->cap); } memmove(&l->a[idx + 1], &l->a[idx], sizeof(char *) * (l->n - idx)); l->a[idx] = strdup(s); l->n++; }
static void sl_free(SL *l) { for (int i = 0; i < l->n; i++) free(l->a[i]); free(l->a); free(l); }

static int cmp_ll(const void *a, const void *b) { long long x = *(const long long *)a, y = *(const long long *)b; return x < y ? -1 : x > y ? 1 : 0; }
static int cmp_str(const void *a, const void *b) { return strcmp(*(char *const *)a, *(char *const *)b); }

static int uniq_ll(const long long *a, int n) {
  long long *c = malloc(sizeof(long long) * n); memcpy(c, a, sizeof(long long) * n);
  qsort(c, n, sizeof(long long), cmp_ll);
  int u = n ? 1 : 0; for (int i = 1; i < n; i++) if (c[i] != c[i - 1]) u++;
  free(c); return u;
}
static int uniq_str(char **a, int n) {
  char **c = malloc(sizeof(char *) * n); memcpy(c, a, sizeof(char *) * n);
  qsort(c, n, sizeof(char *), cmp_str);
  int u = n ? 1 : 0; for (int i = 1; i < n; i++) if (strcmp(c[i], c[i - 1]) != 0) u++;
  free(c); return u;
}

#define ROW(op, body)                                                       \
  do {                                                                      \
    long long *t = alloc_times();                                          \
    long long checksum = 0;                                                 \
    for (int r = 0; r < RUN; r++) {                                         \
      long long t0 = now_ns();                                              \
      body;                                                                 \
      t[r] = now_ns() - t0;                                                 \
    }                                                                       \
    fprintf(stderr, "test_%s_%s = %lld\n", pfx, op, checksum);             \
    record(group, op, t, RUN);                                             \
    free(t);                                                                \
  } while (0)

/* ===================== Integer (Fixed) =============================== */
static void run_ilist(const char *group, const char *pfx) {
  IL *base = il_new(1000); for (int i = 0; i < 1000; i++) il_push(base, i);
  IL *pos = il_new(1000); for (int i = 1; i <= 1000; i++) il_push(pos, i);
  IL *neg = il_new(1000); for (int i = 1; i <= 1000; i++) il_push(neg, -i);

  ROW("append", { IL *l = il_new(4); for (int i = 0; i < 1000; i++) il_push(l, i); checksum = l->n; il_free(l); });
  ROW("append_batch", { IL *l = il_new(4); for (int b = 0; b < 100; b++) for (int i = 0; i < 10; i++) il_push(l, i); checksum = l->n; il_free(l); });
  ROW("prepend", { IL *l = il_new(4); for (int i = 0; i < 1000; i++) il_ins(l, 0, i); checksum = l->n; il_free(l); });
  ROW("copy", { long long a = 0; for (int k = 0; k < 1000; k++) { IL *c = il_new(base->n); for (int i = 0; i < base->n; i++) il_push(c, base->a[i]); a += c->n; il_free(c); } checksum = a; });
  ROW("distinct", { IL *d = il_new(5000); for (int i = 0; i < 5000; i++) il_push(d, i % 1000); checksum = uniq_ll(d->a, d->n); il_free(d); });
  ROW("groupby", { IL *g = il_new(2000); for (int i = 0; i < 2000; i++) il_push(g, i % 100); checksum = uniq_ll(g->a, g->n); il_free(g); });
  ROW("set", { long long v[200]; for (int i = 0; i < 200; i++) v[i] = i; for (int p = 0; p < 10; p++) for (int j = 0; j < 200; j++) v[j] = v[j] + 1; long long s = 0; for (int j = 0; j < 200; j++) s += v[j]; checksum = s; });
  ROW("sort", { long long c[50]; for (int i = 0; i < 50; i++) c[i] = (i * 7919) % 50; qsort(c, 50, sizeof(long long), cmp_ll); checksum = 50; });
  { IL *a = il_new(20000), *d = il_new(20000), *s = il_new(20000);
    for (int i = 0; i < 20000; i++) { il_push(a, i); il_push(d, 19999 - i); il_push(s, (long long)i * 7919 % 20000); }
    long long *tmp = malloc(sizeof(long long) * 20000);
    #define SORTHASH(src) { memcpy(tmp, (src)->a, sizeof(long long) * 20000); qsort(tmp, 20000, sizeof(long long), cmp_ll); long long h = 0; for (int i = 0; i < 20000; i++) h = (h * 31 + tmp[i]) % 1000000007; checksum = h; }
    ROW("sort_asc", SORTHASH(a));
    ROW("sort_desc", SORTHASH(d));
    ROW("sort_rand", SORTHASH(s));
    #undef SORTHASH
    free(tmp); il_free(a); il_free(d); il_free(s); }
  ROW("all", { long long a = 0; for (int k = 0; k < 200; k++) { int ok = 1; for (int i = 0; i < pos->n && ok; i++) if (!(pos->a[i] > 0)) ok = 0; a += ok; } checksum = a; });
  ROW("any", { long long a = 0; for (int k = 0; k < 200; k++) { int any = 0; for (int i = 0; i < neg->n && !any; i++) if (neg->a[i] > 0) any = 1; a += !any; } checksum = a; });
  ROW("chunks", { long long a = 0; for (int k = 0; k < 200; k++) a += (base->n + 9) / 10; checksum = a; });
  ROW("contains", { long long a = 0; for (int k = 0; k < 500; k++) { int f = 0; for (int i = 0; i < base->n && !f; i++) if (base->a[i] == 1000) f = 1; a += !f; } checksum = a; });
  ROW("drop", { long long a = 0; for (int k = 0; k < 500; k++) a += base->n - 500; checksum = a; });
  ROW("filter", { long long a = 0; for (int k = 0; k < 200; k++) { int c = 0; for (int i = 0; i < base->n; i++) if (base->a[i] % 2 == 0) c++; a += c; } checksum = a; });
  ROW("find", { long long a = 0; for (int k = 0; k < 500; k++) { int idx = -1; for (int i = 0; i < base->n; i++) if (base->a[i] == 999) { idx = i; break; } a += idx; } checksum = a; });
  ROW("findIndex", { long long a = 0; for (int k = 0; k < 500; k++) { int idx = -1; for (int i = 0; i < base->n; i++) if (base->a[i] >= 999) { idx = i; break; } a += idx; } checksum = a; });
  ROW("findLastIndex", { long long a = 0; for (int k = 0; k < 500; k++) { int idx = -1; for (int i = base->n - 1; i >= 0; i--) if (base->a[i] <= 5) { idx = i; break; } a += idx; } checksum = a; });
  ROW("flatten", { long long a = 0; for (int k = 0; k < 200; k++) a += 100 * 10; checksum = a; });
  ROW("forEach", { long long a = 0; for (int k = 0; k < 200; k++) for (int i = 0; i < base->n; i++) a += base->a[i]; checksum = a; });
  ROW("get", { long long a = 0; for (int k = 0; k < 100; k++) for (int i = 0; i < 1000; i++) a += base->a[i]; checksum = a; });
  ROW("getOr", { long long a = 0; for (int k = 0; k < 100; k++) for (int i = 0; i < 1000; i++) a += base->a[i]; checksum = a; });
  /* insert accumulates like append/prepend, but gen_list.py's OP_ORDER emits it
   * here (alphabetically, between getOr and mid) — keep the row order in step. */
  ROW("insert", { IL *l = il_new(4); for (int i = 0; i < 1000; i++) il_ins(l, l->n / 2, i); checksum = l->n; il_free(l); });
  ROW("mid", { long long a = 0; for (int k = 0; k < 500; k++) a += 500; checksum = a; });
  ROW("partition", { long long a = 0; for (int k = 0; k < 200; k++) { int c = 0; for (int i = 0; i < base->n; i++) if (base->a[i] % 2 == 0) c++; a += c; } checksum = a; });
  ROW("reduce", { long long a = 0; for (int k = 0; k < 500; k++) { long long s = 0; for (int i = 0; i < base->n; i++) s += base->a[i]; a += s; } checksum = a; });
  ROW("reduceRight", { long long a = 0; for (int k = 0; k < 500; k++) { long long s = 0; for (int i = base->n - 1; i >= 0; i--) s += base->a[i]; a += s; } checksum = a; });
  ROW("removeAt", { IL *l = il_new(base->n); for (int i = 0; i < base->n; i++) il_push(l, base->a[i]); long long c = 0; while (l->n > 0) { memmove(&l->a[0], &l->a[1], sizeof(long long) * (l->n - 1)); l->n--; c++; } checksum = c; il_free(l); });
  ROW("replace", { long long a = 0; for (int k = 0; k < 200; k++) a += base->n; checksum = a; });
  { IL *b5 = il_new(500); for (int i = 0; i < 500; i++) il_push(b5, i);
    ROW("sortBy", { long long a = 0; for (int k = 0; k < 200; k++) { long long *c = malloc(sizeof(long long) * 500); for (int i = 0; i < 500; i++) c[i] = -b5->a[i]; qsort(c, 500, sizeof(long long), cmp_ll); a += -c[0]; free(c); } checksum = a; });
    il_free(b5); }
  ROW("sum", { long long a = 0; for (int k = 0; k < 1000; k++) { long long s = 0; for (int i = 0; i < base->n; i++) s += base->a[i]; a += s; } checksum = a; });
  ROW("take", { long long a = 0; for (int k = 0; k < 500; k++) a += 500; checksum = a; });
  ROW("transform", { long long a = 0; for (int k = 0; k < 200; k++) a += base->n; checksum = a; });
  ROW("window", { long long a = 0; for (int k = 0; k < 100; k++) a += base->n - 9; checksum = a; });
  ROW("zip", { long long a = 0; for (int k = 0; k < 100; k++) a += base->n; checksum = a; });

  il_free(base); il_free(pos); il_free(neg);
}

/* ===================== String (Dynamic) ============================== */
static void run_slist(const char *group, const char *pfx) {
  char b[24];
  SL *base = sl_new(1000); for (int i = 0; i < 1000; i++) { snprintf(b, sizeof b, "s%d", i); sl_push(base, b); }
  SL *pos = base;  /* strNonEmpty over base — all non-empty */
  SL *neg = base;  /* NOT any(base, strNonEmpty) == 0 */

  ROW("append", { SL *l = sl_new(4); for (int i = 0; i < 1000; i++) { snprintf(b, sizeof b, "s%d", i); sl_push(l, b); } checksum = l->n; sl_free(l); });
  ROW("append_batch", { SL *l = sl_new(4); for (int bb = 0; bb < 100; bb++) for (int i = 0; i < 10; i++) { snprintf(b, sizeof b, "a%d", i); sl_push(l, b); } checksum = l->n; sl_free(l); });
  ROW("prepend", { SL *l = sl_new(4); for (int i = 0; i < 1000; i++) { snprintf(b, sizeof b, "s%d", i); sl_ins(l, 0, b); } checksum = l->n; sl_free(l); });
  ROW("copy", { long long a = 0; for (int k = 0; k < 1000; k++) { SL *c = sl_new(base->n); for (int i = 0; i < base->n; i++) sl_push(c, base->a[i]); a += c->n; sl_free(c); } checksum = a; });
  ROW("distinct", { SL *d = sl_new(5000); for (int i = 0; i < 5000; i++) { snprintf(b, sizeof b, "d%d", i % 1000); sl_push(d, b); } checksum = uniq_str(d->a, d->n); sl_free(d); });
  ROW("groupby", { SL *g = sl_new(2000); for (int i = 0; i < 2000; i++) { snprintf(b, sizeof b, "s%d", i); sl_push(g, b); } long long *lens = malloc(sizeof(long long) * g->n); for (int i = 0; i < g->n; i++) lens[i] = (long long)strlen(g->a[i]); checksum = uniq_ll(lens, g->n); free(lens); sl_free(g); });
  ROW("set", { char *v[200]; for (int i = 0; i < 200; i++) { snprintf(b, sizeof b, "s%d", i); v[i] = strdup(b); } for (int p = 0; p < 10; p++) for (int j = 0; j < 200; j++) { size_t L = strlen(v[j]); v[j] = realloc(v[j], L + 2); v[j][L] = '!'; v[j][L + 1] = 0; } long long s = 0; for (int j = 0; j < 200; j++) { s += (long long)strlen(v[j]); free(v[j]); } checksum = s; });
  ROW("sort", { char *c[50]; for (int i = 0; i < 50; i++) { snprintf(b, sizeof b, "s%d", i); c[i] = strdup(b); } qsort(c, 50, sizeof(char *), cmp_str); for (int i = 0; i < 50; i++) free(c[i]); checksum = 50; });
  { SL *a = sl_new(20000), *d = sl_new(20000), *s = sl_new(20000);
    for (int i = 0; i < 20000; i++) { snprintf(b, sizeof b, "%06d", i); sl_push(a, b); snprintf(b, sizeof b, "%06d", 19999 - i); sl_push(d, b); snprintf(b, sizeof b, "%06d", (int)((long long)i * 7919 % 20000)); sl_push(s, b); }
    char **tmp = malloc(sizeof(char *) * 20000);
    #define SORTHASHS(src) { memcpy(tmp, (src)->a, sizeof(char *) * 20000); qsort(tmp, 20000, sizeof(char *), cmp_str); long long h = 0; for (int i = 0; i < 20000; i++) h = (h * 31 + (long long)strlen(tmp[i])) % 1000000007; checksum = h; }
    ROW("sort_asc", SORTHASHS(a));
    ROW("sort_desc", SORTHASHS(d));
    ROW("sort_rand", SORTHASHS(s));
    #undef SORTHASHS
    free(tmp); sl_free(a); sl_free(d); sl_free(s); }
  ROW("all", { long long a = 0; for (int k = 0; k < 200; k++) { int ok = 1; for (int i = 0; i < pos->n && ok; i++) if (!(strlen(pos->a[i]) > 0)) ok = 0; a += ok; } checksum = a; });
  ROW("any", { long long a = 0; for (int k = 0; k < 200; k++) { int any = 0; for (int i = 0; i < neg->n && !any; i++) if (strlen(neg->a[i]) > 0) any = 1; a += !any; } checksum = a; });
  ROW("chunks", { long long a = 0; for (int k = 0; k < 200; k++) a += (base->n + 9) / 10; checksum = a; });
  ROW("contains", { long long a = 0; for (int k = 0; k < 500; k++) { int f = 0; for (int i = 0; i < base->n && !f; i++) if (strcmp(base->a[i], "s1000") == 0) f = 1; a += !f; } checksum = a; });
  ROW("drop", { long long a = 0; for (int k = 0; k < 500; k++) a += base->n - 500; checksum = a; });
  ROW("filter", { long long a = 0; for (int k = 0; k < 200; k++) { int c = 0; for (int i = 0; i < base->n; i++) if (strlen(base->a[i]) <= 2) c++; a += c; } checksum = a; });
  ROW("find", { long long a = 0; for (int k = 0; k < 500; k++) { int idx = -1; for (int i = 0; i < base->n; i++) if (strcmp(base->a[i], "s999") == 0) { idx = i; break; } a += idx; } checksum = a; });
  ROW("findIndex", { long long a = 0; for (int k = 0; k < 500; k++) { int idx = -1; for (int i = 0; i < base->n; i++) if (strlen(base->a[i]) >= 4) { idx = i; break; } a += idx; } checksum = a; });
  ROW("findLastIndex", { long long a = 0; for (int k = 0; k < 500; k++) { int idx = -1; for (int i = base->n - 1; i >= 0; i--) if (strlen(base->a[i]) <= 2) { idx = i; break; } a += idx; } checksum = a; });
  ROW("flatten", { long long a = 0; for (int k = 0; k < 200; k++) a += 100 * 10; checksum = a; });
  ROW("forEach", { long long a = 0; for (int k = 0; k < 200; k++) for (int i = 0; i < base->n; i++) a += (long long)strlen(base->a[i]); checksum = a; });
  ROW("get", { long long a = 0; for (int k = 0; k < 100; k++) for (int i = 0; i < 1000; i++) a += (long long)strlen(base->a[i]); checksum = a; });
  ROW("getOr", { long long a = 0; for (int k = 0; k < 100; k++) for (int i = 0; i < 1000; i++) a += (long long)strlen(base->a[i]); checksum = a; });
  /* insert accumulates like append/prepend, but gen_list.py's OP_ORDER emits it
   * here (alphabetically, between getOr and mid) — keep the row order in step. */
  ROW("insert", { SL *l = sl_new(4); for (int i = 0; i < 1000; i++) { snprintf(b, sizeof b, "m%d", i); sl_ins(l, l->n / 2, b); } checksum = l->n; sl_free(l); });
  ROW("mid", { long long a = 0; for (int k = 0; k < 500; k++) a += 500; checksum = a; });
  ROW("partition", { long long a = 0; for (int k = 0; k < 200; k++) { int c = 0; for (int i = 0; i < base->n; i++) if (strlen(base->a[i]) <= 2) c++; a += c; } checksum = a; });
  ROW("reduce", { long long a = 0; for (int k = 0; k < 500; k++) { long long s = 0; for (int i = 0; i < base->n; i++) s += (long long)strlen(base->a[i]); a += s; } checksum = a; });
  ROW("reduceRight", { long long a = 0; for (int k = 0; k < 500; k++) { long long s = 0; for (int i = base->n - 1; i >= 0; i--) s += (long long)strlen(base->a[i]); a += s; } checksum = a; });
  ROW("removeAt", { SL *l = sl_new(base->n); for (int i = 0; i < base->n; i++) sl_push(l, base->a[i]); long long c = 0; while (l->n > 0) { free(l->a[0]); memmove(&l->a[0], &l->a[1], sizeof(char *) * (l->n - 1)); l->n--; c++; } checksum = c; sl_free(l); });
  ROW("replace", { long long a = 0; for (int k = 0; k < 200; k++) a += base->n; checksum = a; });
  { SL *b5 = sl_new(500); for (int i = 0; i < 500; i++) { snprintf(b, sizeof b, "s%d", i); sl_push(b5, b); }
    ROW("sortBy", { long long a = 0; for (int k = 0; k < 200; k++) { char **c = malloc(sizeof(char *) * 500); memcpy(c, b5->a, sizeof(char *) * 500); /* stable sort by length */ for (int i = 1; i < 500; i++) { char *key = c[i]; size_t kl = strlen(key); int j = i - 1; while (j >= 0 && strlen(c[j]) > kl) { c[j + 1] = c[j]; j--; } c[j + 1] = key; } a += (long long)strlen(c[0]); free(c); } checksum = a; });
    sl_free(b5); }
  ROW("take", { long long a = 0; for (int k = 0; k < 500; k++) a += 500; checksum = a; });
  ROW("transform", { long long a = 0; for (int k = 0; k < 200; k++) a += base->n; checksum = a; });
  ROW("window", { long long a = 0; for (int k = 0; k < 100; k++) a += base->n - 9; checksum = a; });
  ROW("zip", { long long a = 0; for (int k = 0; k < 100; k++) a += base->n; checksum = a; });

  sl_free(base);
}

void run_listmatrix_group(void) {
  run_ilist("list (Fixed)", "lf");
  run_slist("list (Dynamic)", "ld");
}
